src/go/rpk/pkg/api/admin/admin.go

```diff
@@ -185,39 +215,112 @@ func (a *AdminAPI) GetLeaderID() (*int, error) {
+		if err == nil {
+			// Success, return the result from this node.
+			return maybeUnmarshalRespInto(method, url, res, into)
+		} else {
+			if i != len(shuffled)-1 {
+				// We will retry on another node, log the error from this node.
+				log.Infof("Request error, trying another node: %s", err.Error())
+			}
+		}
 	}
-	return maybeUnmarshalRespInto(method, url, res, into)
+
+	// Fall through: all nodes failed.
+	return err
 }
 
+// General purpose backoff, includes 503s and other errors
+const retryBackoff = time.Duration(1500) * time.Millisecond
+
+// When there is no leader, we wait long enough for an election to complete
+const noLeaderBackoff = time.Duration(1500) * time.Millisecond
+
+// When there is a stale leader, we might have to wait long enough for
+// an election to start *and* for the resulting eleciton to complete
+const staleLeaderBackoff = time.Duration(9000) * time.Millisecond
+
 // sendToLeader sends a single request to the leader of the Admin API for Redpanda >= 21.11.1
 // otherwise, it broadcasts the request
 func (a *AdminAPI) sendToLeader(
 	method, path string, body, into interface{},
 ) error {
 	// If there's only one broker, let's just send the request to it
 	if len(a.urls) == 1 {
-		return a.sendOne(method, path, body, into)
-	}
-	leaderID, err := a.GetLeaderID()
-	if err != nil {
-		return err
-	}
-	url, err := a.brokerIDToURL(*leaderID)
-	// if it's not possible to map the leaderID to a broker URL -> broadcast
-	if err != nil {
-		return a.sendAll(method, path, body, into)
+		return a.sendOne(method, path, body, into, true)
+	}
+
+	retries := 3
+	var leaderID *int
+	var leaderURL string
+	for leaderID == nil || leaderURL == "" {
+		var err error
+		leaderID, err = a.GetLeaderID()
+		if _, ok := err.(*ErrNoAdminAPILeader); ok {
+			// No leader?  We might have contacted a recently-started node
+			// who doesn't know yet, or there might be an election pending,
+			// or there might be no quorum.  In any case, retry in the hopes
+			// the cluster will get out of this state.
+			retries -= 1
+			if retries == 0 {
+				return err
+			}
+			time.Sleep(noLeaderBackoff)
+		} else if err != nil {
+			// Unexpected error, do not retry promptly.
+			return err
+		} else {
+			// Got a leader ID, check if it's resolvable
+			leaderURL, err = a.brokerIDToURL(*leaderID)
+			if err != nil && len(a.brokerIdToUrls) == 0 {
+				// Could not map any IDs: probably this is an old redpanda
+				// with no node_config endpoint.  Fall back to broadcast.
+				return a.sendAll(method, path, body, into)
+			} else if err != nil {
+				// Have ID mapping for some nodes but not the one that is
+				// allegedly the leader.  This leader ID is probably stale,
+				// e.g. if it just died a moment ago.  Try again.  This is
+				// a long timeout, because it's the sum of the time for nodes
+				// to start an election, followed by the worst cast number of
+				// election rounds
```

Comment:
Would it be better to retry sooner and more often, rather than to assume the worst case and sleep 9s? If I run a command at my terminal such as dump my local config and it hangs for 9s, I'd probably ctrl+c after three.

Also... is this case extremely rare?

Reply:
It's as rare as a cluster node failure is -- pretty rare on real systems, although also likely to happen at exactly the times the user might be using rpk to poke at the system while doing maintenance (and perhaps restarting nodes).  Way more frequent in our testing of redpanda, where we inject node failures.

Reply:
It still may be better to have shorter retries here -- AFAICT this is essentially assuming the worst case: rpk hit the cluster immediately once election began, and then assumes the worst case for how long election can take. On average (for this rare case) I'd expect rpk to hit the cluster in the middle or later of this cycle, so a 4.5s retry, and AFAICT, a 1s retry isn't really detrimental.

It may speed up tests to retry more often, but perhaps we can wait for experience reports.